README.md
# balena-fin-firmata-flash

 This is a balena application template that has a co-processor service that allows for flashing of the balenaFin co-processor with the latest version of the [balenaFin Firmata](https://github.com/balena-io/balena-fin-coprocessor-firmata). For more information about Firmata, check out the official [repo](https://github.com/firmata/protocol).

**NOTE:** while having this interface exposed to the outside is useful for development (you can consume this interface from your laptop in order to flash and test your balenaFin) we highly suggest to change the docker-compose configuration for the co-processor service ( `network_mode: host` ) once your own business logic has this interface implemented so that only the services you deploy on the device will be able to access it.

> :warning: **Flashing the balenaFin v1.1 will result in a device reboot**: Be aware of this when flashing!

# balenaFin device tree overlay setup

in order to allow the Raspberry Pi Compute Module to communicate via UART to the co-processor, you need to set the following configuration variable on your device (or app-wide)

`BALENA_HOST_CONFIG_dtoverlay` = `"balena-fin","uart1,txd1_pin=32,rxd1_pin=33"`

### co-processor REST interface (port 1337)

default port: `1337`

##### flash firmware

_POST_ `/v1/flash/:firmware`

flashes the given firmware name from the service fs ( ie `firmata-balena-0.0.2.hex` ). The co-processor service has a volume (`/data`) that you can add to other service and use as a way to share different firmwares to flash. on balenaFin v1.1 and above, this action will trigger a reboot.

##### sleep

_POST_ `/v1/sleep/:delay/:timeout`

triggers the balenaFin power saving mode.
* `delay` (integer) is the amount of seconds the co-processor will wait before shutting down the linux side (CM3/CM3+)
* `timeout` (integer) is the amount of seconds the co-processor will keep the linux side (CM3/CM3+) shut down before bringing it back up. There is a limit of 97 years (3,058,992,000 seconds) as the max value the co-processor can handle
* you can override update checks that would otherwise prevent the sleep from being triggered passing `force` in the body of the request: `{"force":1}`

##### set pin

_POST_ `/v1/setpin/:pin/:state`

set digital pin state on the coprocessor header.
* `pin` (integer) is the `Expansion Header` pin numbering as shown in the image below.
* `state` (integer) is either 1 (on) or 0 (off)

Reference for the pin numbers relating to the GPIO `Expansion Header` can be found in the [firmata source](https://github.com/balena-io/balena-fin-coprocessor-firmata#firmata-pin-map) repository.

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v2.2.0
## (2021-04-09)

* Removed JTAG pins as they are un-used [Alex Bucknall]

# v2.1.3
## (2021-04-05)

* Tweaked docs to highlight that a reboot will occur when flashing the v1.1 [Alex Bucknall]

# v2.1.2
## (2020-11-17)

* enable with rev 11 [rcooke-warwick]

# v2.1.1
## (2020-10-06)

* Dependabot bump node-fetch [Alex Bucknall]

# v2.1.0
## (2020-10-06)

* allow sleep to be forced [curcuz]

# v2.0.1
## (2020-07-28)

* Cleans up GPIO reference [Alex Bucknall]

# v2.0.0
## (2020-06-26)

* target major node version [curcuz]

# v1.1.1
## (2020-05-19)

* Added firmata firmware v2.0.1 [Alex Bucknall]

# v1.1.0
## (2020-04-22)

* Use impl version to detect if automatic flashing is needed [Roman Mazur]

# v1.0.1
## (2020-04-21)

* Run balena build with balenaCI on PR [Roman Mazur]

# v1.0.0
## (2020-04-21)

* Corrected docs endpoint [Alex Bucknall]
* Add version file [Roman Mazur]
* Add package-lock file [Roman Mazur]
* Added lockfile to coprocessor flashing [Alex Bucknall]
* Allow automatic flashing [Roman Mazur]
* add balena-fin-sdk for hw rev detection [curcuz]


app/main.js
const readline = require('readline');
const Firmata = require("firmata");
const board = new Firmata(process.env.FIRMATA_DEVICE || "/dev/ttyS0");

readline.emitKeypressEvents(process.stdin);
process.stdin.setRawMode(true);

// board.settings.skipCapabilities = true;
console.log("Checking Firmata version...")

board.on("queryfirmware", () => {
    // Encoded terminal text colors, i.e. green is \x1b[32m
    console.log('\x1b[32m%s\x1b[0m', board.firmware.name +"  ✔ firmware name");
    console.log('\x1b[32m%s\x1b[0m', board.firmware.version.major + "." + board.firmware.version.minor + "              ✔ firmata version")
});

board.on("ready", function() {
  const LED = 14;

  const DIGITAL_OUT = 2;
  let state = 0;
  let DIGITAL_PASS = false;

  const ANALOG_OUT = 4;
  const ANALOG_IN = 3; 
  let ANALOG_PASS = false;
  const analogs = [ANALOG_IN];
  let i = 0;

  const POWER_DELAY = 5;
  const POWER_SLEEP = 20;

  console.log('\x1b[32m%s\x1b[0m', "balenaFin        ✔ ready");
  console.log('\x1b[35m%s\x1b[0m', "Press the 's' key to enter sleep test. Warning this will power down your balenaFin and the ssh connection will terminate!");
  console.log('\x1b[34m%s\x1b[0m', "Press the 'c' key to terminate the test.");
  console.log("Starting DIGITAL I/O check...");

  this.pinMode(LED, board.MODES.OUTPUT);
  this.pinMode(DIGITAL_OUT, board.MODES.OUTPUT);
  this.digitalWrite(DIGITAL_OUT, 1);

  const states = {
    1 : 0
  };

  Object.keys(states).forEach(function(pin) {
    pin = +pin;
    this.pinMode(pin, board.MODES.INPUT);
    this.digitalRead(pin, function(value) {
      console.log("DIGITAL_IN | Pin %d | Value %d", pin, value);
      if(value == 1){DIGITAL_PASS = true;}
    });
  }, this);

  process.stdin.on('keypress', (str, key) => {
    if(key.sequence === '\u0073') {
      console.log('\x1b[31m%s\x1b[0m', "Entering Sleep Test...");
      console.log('\x1b[31m%s\x1b[0m', `Fin will power down in ${POWER_DELAY} seconds and remain powered down for ${POWER_SLEEP} seconds...`);
      this.sysexCommand(configSleep(POWER_DELAY,POWER_SLEEP));
    }
    if(key.sequence === '\u0063') {
      console.log('\x1b[31m%s\x1b[0m', "Terminating test and resetting Firmata");
      board.reset();
      process.exit();
    }
  });

  console.log("Starting ANALOG I/O check...");

  this.pinMode(ANALOG_OUT, board.MODES.PWM);

  analogs.forEach(function(pin) {
    pin = +pin;
    this.pinMode(ANALOG_IN, board.MODES.ANALOG);
    this.analogRead(ANALOG_IN, function(value) {
      if(value > 2000){
        ANALOG_PASS = true;
      };
      // loops through duty cycles of PWM
      i = 10 + i;
      console.log("ANALOG_IN  | Pin %d | Value %d", ANALOG_IN, value);
      this.analogWrite(ANALOG_OUT, i);
      if(i >= 100){ i = 0}
    });
  }, this);



  setInterval(() => {
    this.digitalWrite(LED, (state ^= 1));
    if(DIGITAL_PASS && ANALOG_PASS){
      console.log('\x1b[32m%s\x1b[0m', "All checks passed ✔");
      process.exit();
    }
  }, 500);
});

process.on('SIGINT', function() {
  console.log("Resetting Firmata...");
  board.reset();
  process.exit();
});

function configSleep(delay,sleep) {
  // we want to represent the input as a 8-bytes array
  var byteArray = [0, 0, 0, 0, 0, 0];

  // pad every 8th bit with 0 for conformance with Firmata protocol
  var sleep_string = sleep.toString(2);
  sleep_string = padData(sleep_string);
  var sleep_data = parseInt(sleep_string, 2);

  var delay_string = delay.toString(2);
  delay_string = padData(delay_string);
  var delay_data = parseInt(delay_string, 2);

  for ( var index = 0; index < byteArray.length; index ++ ) {
      var byte = sleep_data & 0xff;
      byteArray[index] = byte;
      sleep_data = (sleep_data - byte) / 256 ;
  }

  // insert balena command
  byteArray.splice(0, 0, 0x0B);
  // insert balena subcommand
  byteArray.splice(1, 0, 0x01);  
  byteArray.splice(2, 0, delay_data);  
  console.log(byteArray)
  return byteArray;
};

function padData(input_string) {
  for (var i = 1; i < 6; i++) {
    input_string = input_string.splice((i*-7)-((i*1)-1), 0, "0");
  };
  return input_string
}

String.prototype.splice = function(idx, rem, str) {
  return this.slice(0, idx) + str + this.slice(idx + Math.abs(rem));
};

co-processor/app/flash.sh
#!/bin/bash

FW=$1
REV=$2

# Makes sure we exit if lock fails.
set -e

echo "acquiring lockfile..."
exec {lock_fd}>/tmp/balena/updates.lock || exit 1
flock -n "$lock_fd" || { echo "ERROR: failed to acquire lockfile." >&2; rm -f /tmp/balena/updates.lock; exit 1; }

echo "opening screen terminal for flashing $FW to balenaFin v$REV"
case $REV in
  09)
    screen -dmS swd_program  "./openocd/openocd-v1.0.sh"
    ;;
  10)
    screen -dmS swd_program  "./openocd/openocd-v1.1.sh"
    ;;
  11)
    screen -dmS swd_program  "./openocd/openocd-v1.1.sh"
    ;;
  *)
    echo "ERROR: unknown balenaFin revision" >&2
    flock -u "$lock_fd"; rm -f /tmp/balena/updates.lock
    exit 1
    ;;
esac
sleep 6
  { sleep 5; echo "reset halt"; echo "program firmware/bootloader.s37"; sleep 5; echo "reset halt"; echo "program firmware/$FW"; echo "reset run"; sleep 10; echo "exit"; echo -e '\x1dclose\x0d'; } | telnet localhost 4444
sleep 5
echo -e "flashing complete\n"
echo "releasing lockfile..."
flock -u "$lock_fd"; rm -f /tmp/balena/updates.lock
echo "closing the openocd process..."
kill $(ps aux | grep '[S]CREEN -dmS swd_program' | awk '{print $2}')


co-processor/app/index.js
#!/bin/env node

const express = require('express');
const compression = require('compression');
const bodyParser = require('body-parser');

const gi = require('node-gtk');
const Fin = gi.require('Fin', '0.2');
const fin = new Fin.Client();
const BALENA_FIN_REVISION = fin.revision;
const SERVER_PORT = parseInt(process.env.SERVER_PORT) || 1337;

const firmata = require(__dirname + '/utils/firmata.js');
const supervisor = require(__dirname + '/utils/supervisor.js');
const Flasher = require(__dirname + '/flasher.js');

const app = express();
const flasher = Flasher(BALENA_FIN_REVISION, supervisor, firmata);

const shutdown = (delay, timeout, force) => {
  if (force) {
    firmata.sleep(parseInt(delay), parseInt(timeout));
        return supervisor.shutdown();
  } else {
    return supervisor.checkForOngoingUpdate()
      .then(() => {
        firmata.sleep(parseInt(delay), parseInt(timeout));
        return supervisor.shutdown();
      })
      .catch(() => { throw new Error('Device is not Idle, likely updating, will not shutdown'); });
  }
  
};

let setPin = function(pin,state) {
  return new Promise((resolve, reject) => {
    supervisor.checkForOngoingUpdate().then((response) => {
      firmata.setPin(parseInt(pin), parseInt(state));
      resolve();
    }).catch((response) => {
      reject("coprocessor is not responding...");
    });
  });
};

const errorHandler = (err, req, res, next) => {
  res.status(500);
  res.render('error', {
    error: err
  });
};

app.use(compression());
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({
  extended: true
}));
app.use(function(req, res, next) {
  res.header("Access-Control-Allow-Origin", "*");
  res.header('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE');
  res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
  next();
});
app.use(errorHandler);

app.post('/v1/flash/:fw', (req, res) => {
  if (!req.params.fw) {
    return res.status(400).send('Bad Request');
  }

  return flasher.flash(req.params.fw)
      .then(() => {
        return res.status(200).send('OK');
      });
});

app.post('/v1/setpin/:pin/:state', (req, res) => {
  if (!req.params.pin || !req.params.state) {
    return res.status(400).send('Bad Request');
  }
  console.log('set ' + req.params.pin + ' to state ' + req.params.state);
  setPin(req.params.pin, req.params.state).then(()=> {
    res.status(200).send('OK');
  }).catch((error) => {
    console.error("device is not responding, check for on-going coprocessor flashing/application updating.");  });
});

app.post('/v1/sleep/:delay/:timeout', (req, res) => {
  const force = (typeof req.body.force === "undefined") ? 0 : parseInt(req.body.force);
  if (!req.params.delay || !req.params.timeout) {
    return res.status(400).send('Bad Request');
  }
  if (parseInt(BALENA_FIN_REVISION) < 10) {
    return res.status(405).send('Feature not available on current hardware revision');
  }

  shutdown(req.params.delay, req.params.timeout, force)
      .then(() => {
        console.error("Sleep command registered, shutting down...");
        res.status(200).send('OK');
      })
      .catch((error) => {
        console.error(`Error registering sleep command: ${error.message}`);
        throw error;
      });
});

app.listen(SERVER_PORT, () => {
  console.log('balenaFin revision', BALENA_FIN_REVISION);
  console.log('server listening on port ' + SERVER_PORT);
});

process.on('SIGINT', () => {
  flasher.stop();
  process.exit();
});

if (process.env.FLASH_ON_START_FILE) {
  const firmwareFile = process.env.FLASH_ON_START_FILE;
  const firmwareMeta = {
    name: process.env.FLASH_ON_START_NAME,
    version: process.env.FLASH_ON_START_VERSION,
  };

  if (firmwareMeta.name && firmwareMeta.version) {
    console.log(`Automatic flashing is configured using ${firmwareFile}.`);
    flasher.flashIfNeeded(firmwareFile, firmwareMeta)
        .then((flashed) => {
          if (!flashed) {
            console.log('Automatic flashing is skipped: the requested firmware is already flashed.');
          }
        })
        .catch(console.error);
  } else {
    console.log('Bad automatic flash configuration: firmware meta data is not fully provided');
  }
}
